code/ui/blocks/src/blocks/Canvas.tsx

```diff
@@ -1,76 +1,66 @@
-import type { FC, ReactElement, ReactNode, ReactNodeArray } from 'react';
-import React, { useContext } from 'react';
+import React, { Children, useContext } from 'react';
+import type { FC, ReactElement, ReactNode } from 'react';
 import type { Framework } from '@storybook/types';
 import type { PreviewProps as PurePreviewProps } from '../components';
 import { Preview as PurePreview, PreviewSkeleton } from '../components';
 import type { DocsContextProps } from './DocsContext';
 import { DocsContext } from './DocsContext';
 import type { SourceContextProps } from './SourceContainer';
 import { SourceContext } from './SourceContainer';
-import { getSourceProps, SourceState } from './Source';
+import { useSourceProps, SourceState } from './Source';
 import { useStories } from './useStory';
 
 export { SourceState };
 
-type CanvasProps = PurePreviewProps & {
+type CanvasProps = Omit<PurePreviewProps, 'isExpanded'> & {
   withSource?: SourceState;
   mdxSource?: string;
 };
 
-const getPreviewProps = (
+const usePreviewProps = (
```

Comment:
The diff for this function may seem like a lot, but it maintains the exact same functionality and logic, it's just a refactoring. The refactoring was sparked by the need for `useSourceProps` hook to be called unconditionally, and so in an attempt to figure the whole thing out I simplified the logic paths.